cluster.go

```diff
@@ -459,13 +474,73 @@ func (c *Cluster) Do(a Action) error {
 	return c.doInner(a, addr, key, false, doAttempts)
 }
 
+func (c *Cluster) getClusterDownSince() int64 {
+	return atomic.LoadInt64(&c.lastClusterdown)
+}
+
+func (c *Cluster) setClusterDown(down bool) (changed bool) {
+	prevVal := atomic.LoadInt64(&c.lastClusterdown)
+	// avoid CAS in the common case were the cluster is up
+	if !down && prevVal == 0 {
+		return false
+	}
+
+	var newVal int64
+	if down {
+		newVal = time.Now().UnixNano() / 1000 / 1000
+		// check for at least a millisecond difference, to avoid unnecessary stores
+		if newVal-prevVal <= 0 {
+			return false
+		}
+	}
+
+	for !atomic.CompareAndSwapInt64(&c.lastClusterdown, prevVal, newVal) {
+		updatedVal := atomic.LoadInt64(&c.lastClusterdown)
+		// if another goroutine changed the state before we could, we can just return
+		// and avoid more CAS operations since the important thing (the state change)
+		// already happened and the difference in newVal and updatedVal will be small
+		// (when the cluster is down)
+		if prevVal == 0 || updatedVal == 0 || newVal == updatedVal {
+			return false
+		}
+		prevVal = updatedVal
+	}
+
+	changed = (prevVal == 0 && newVal != 0) || (prevVal != 0 && newVal == 0)
+
+	if changed && c.co.ct.StateChange != nil {
+		// we assume that this returns quick. we could also do this
```

Comment:
this comment might be out of date, since there's no lock anymore

Reply:
Done